### ISSUE TITLE
Best way to create custom dataset with real data

### ISSUE THREAD

**Issue by Zrrr1997** [kirumang/Pix2Pose#18]
Hi Kiru,

### My setup
I am using Pix2Pose for a project where I am estimating the pose of a moving car from a static camera. The scene is very simple: the camera is **static** and the car is moving only as a forward and backward **translation** (no turns).

I am using the official CAD model for the car for the pose estimation. 

### My question to you
I want to use real training data to avoid the "reality gap" from synthetic data. I am using markers attached to the windshield of the car to estimate the ground truth pose with respect to them. However this has two issues:
- The pose of the markers is slightly noisy, so we can expect the Pix2Pose-GAN to inherit this noise as well
- The pose is with respect to the marker, so I have to manually adjust the CAD-Modell coordinate system to be in sync with the marker pose, which will never be perfect

What is a good general way to generate **ground truth** 6D pose data? Is it possible to achieve a good offline pose accuracy with just one camera or should I include more to reduce the variance? Or if I want good poses, should I even turn to hardcore equipment, such as LiDAR scanners? I would like to achieve a GT-Accuracy of under 1.5cm.

Please let me know if you know any good techniques to achieve my task or refer me to some literature. In the end, the Pix2Pose model cannot be better than the training data you give it, right? :)

**Comment by kirumang** [comment]
Hi, sorry for the late answer. 

First of all, it is true that Pix2pose cannot be better than the accuracy of the training data, and, if the GT poses and training images contain some biases, I would expect that Pix2Pose produces biased estimations. But, if synthetic data (w/ precise annotations for sure) is used together with "noisy" real data, the biases may be fixed by synthetic images. But, not sure without really trying it.  On the other hand, if you are going to use RGB images (not RGB-D or LiDAR inputs) and have good 3D models of target objects(car in your case), it would be better to check some methods that refine the poses purely on RGB by comparing the renderings of the target object. e.g., [CosyPose@ECCV20](https://arxiv.org/abs/2008.08465). Thus, the pose could be adjusted even the initial estimate is not accurate due to noisy pose annotations. 

For the ground-truth of 6D data, it is always better to have depth measurements in order to derive better annotations even though only RGB images are going to be used for training/inference. Like most 6D pose datasets (usually for small objects on tables),  it is also better to build a multi-camera setup (i.e. multiple cameras with known extrinsic) or multi-view setup (i.e. a camera with known camera poses/camera pose tracking). With these multi-view or multi-camera setups, you could potentially reconstruct a 3D model (in this case, camera poses can be more precisely derived during the reconstruction process with SfM or whatever reconstruction algorithm/SLAM methods) of the target scene(or objects) where you can precisely annotate 3D poses of objects and use the relative camera poses to annotate every single frame, which in general provides reliable poses for the small objects in my experience.  

I would love to point out this paper from TU Munich, [HomebrewedDB-Kaskman et al.](https://arxiv.org/abs/1904.03167), which provides very good 6D pose annotations with an automatic process. 

I hope this helps!

**Comment by HamilkarBrq** [comment]
Hey Kiru,
you suggest to use BlenderProc for the creation of training data for a custom dataset in Issue #12 while you recommend using the HomebrewedDB software for the annotations here. My question is if are both tools eligible for the creation of a custom dataset + 6D pose annotations?
Cheers

**Comment by kirumang** [comment]
> Hey Kiru,
> you suggest to use BlenderProc for the creation of training data for a custom dataset in Issue #12 while you recommend using the HomebrewedDB software for the annotations here. My question is if are both tools eligible for the creation of a custom dataset + 6D pose annotations?
> Cheers

Hi, 

As far as I understood, the original question here was how to create a custom dataset using real images or how to annotate real images efficiently. And, the answer to issue #12 was general ways of preparing a custom dataset, and if there is no real image for training and if textured 3D models are available, I would recommend using BlenderProc for generating synthetic images. What I proposed with HomeBrewedDB was to use a stream of real images (a video) where you can build a 3D reconstruction. By this, you could obtain real training images by propagating pose annotations in the 3D reconstruction to each frame. 

In summary, 
- For the creation of synthetic images using textured 3D models: using BlenderProc (instead of the standard OpenGL)
- Creation of real images while moving cameras (if possible, with markers): have a look at how the HomeBrewedDB annotates their dataset. 

Best,

**Comment by HamilkarBrq** [comment]
Sry my bad, I misread that detail. Thanks for the clarification!
